### PR TITLE
Linux support for the OpenCL SDK package

### PR DESCRIPTION
**Pull request by MathiasMagnus**
I have added Linux support for the OpenCL SDK.

There are two questions I would like to ask before my quest on updating my other packages as well:

1. How does one handle the case when a package typically cannot be compiled to the target platform with the set of restrictions the library imposes?

Although it is said that the OpenCL ICD (the library one links to, that in turn will load the vendor implementations) may not be compiled as a static library (there's never a rationale, [it's just stated](https://forums.khronos.org/showthread.php/8717-build-static-OpenCL-library), perhaps @benjamincoquelle can clarify), I have not encountered any problems neither on Windows nor Linux. I compile it as a static lib, link to it, run my app, it works fine with Intel, AMD, POCL runtimes. It works. On Windows there's an annoying linker error (LIBCMT) I could not yet get rid of, so for the time being, I kept it the way it was, meaning even when static linkage is requested, it forcefully falls back to dynamic linking to the vendor OpenCL.dll. On Linux, I could do the same, but then there would be no artifacts to link to. I could link to a libOpenCL.so, but because the Github master commit is most certainly ahead of an LTS library per say, I would not like to use cutting-edge headers and .so, potentially exposing features not available when using the system library. Instead, I chose to create a static library anyway and wait until someone reports an error. I have not found one until now.

2. What should my portfile hold if the package is pre-installed on a platform or is only available through its native package manager?

I have not tested this (I don't have direct access to a macOS machine, but will ask my colleague sometime soon) to lend me his machine to try macOS support there. The OpenCL SDK is an integral part of the system and in this case installing my package should silently (or with a message) succeed unconditionally, because there is really no artifact to produce. System paths hold the headers and library supported by the drivers shipping with the system. What should I add to my package port file is that's the case? Sandwich the entire things with an `if(VCPKG_CMAKE_SYSTEM_NAME STREQUAL "Darwin")` and be done with it? Will package checkers succeed without any artifacts?

Following this update, I'll go on to updating clFFT, clBLAS packages accordingly. I'll try adding [clBLAST](https://github.com/CNugteren/CLBlast) and [SYCL-BLAS](https://github.com/codeplaysoftware/sycl-blas) as more recent competing alternatives.

### CODE REVIEW

**Comment by MathiasMagnus** [comment]
Let me know if I can be of any assistance with this. I am really trying to be a good citizen here.

**Comment by NancyLi1013** [comment]
Hi @MathiasMagnus, here are the test results from the current CI system:

x64-windows|test|notes
|-----------|----|-----
|clfft      |Fail|**Regression**
|clblast    |Fail|**Regression**
|clblas     |Fail|**Regression**

x64-windows-static had changes from the baseline results

x64-osx had changes from the baseline results

|arm64-windows|test|notes
|-------------|----|-----
|clfft        |Fail|Previously skipped, not a regression
|clblast      |Fail|Previously skipped, not a regression
|clblas       |Fail|Previously skipped, not a regression
|opencl       |Pass|

|x86-windows|test|notes
|-----------|----|-----
|clfft      |Fail|**Regression**
|clblast    |Fail|**Regression**
|clblas     |Fail|**Regression**

[failureLogs.zip](https://github.com/microsoft/vcpkg/files/3216143/failureLogs.zip)

**Comment by MathiasMagnus** [comment]
Thank you @NancyLi1013, that is helpful info indeed. I've been wondering why there are no CI tests for packages, but it seems the team got around to implementing one. Is the `vsts.yaml` file also available so I can check the output in my working branches? Or should most work be done in WIP PR-s so we can see CI output as we advance a given feature? What is the accepted modus operandi?

**Comment by cbezault** [comment]
@MathiasMagnus Contributors generally do most of their work in WIP PRs so that they can see the CI results and we can give continuous feedback.

**Comment by cbezault** [comment]
@MathiasMagnus the ci has now moved to a public facing site.

**Review by cbezault — CHANGES_REQUESTED** [review]

**Comment by cbezault** [comment]
Do you know what the effect of removing the WDK dependency is? If it's important we can just have a message that tells the user they need to install it.

File: ports/opencl/portfile.cmake [hunk 1]
```diff
@@ -1,18 +1,24 @@
 include(vcpkg_common_functions)
 
-vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)
+file(COPY
+        ${CMAKE_CURRENT_LIST_DIR}/vcpkg-cmake-wrapper.cmake
+    DESTINATION
+        ${CURRENT_PACKAGES_DIR}/share/${PORT}
+)
+
+#vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)
```
[the remark this answers is not in the document]
Comment (MathiasMagnus) [reply ports/opencl/portfile.cmake]: Tick

File: ports/opencl/portfile.cmake [hunk 2]
```diff
@@ -49,34 +55,66 @@ message(STATUS "Generating OpenCL C++ headers done")
...
     OUT_SOURCE_PATH SOURCE_PATH
     REPO KhronosGroup/OpenCL-ICD-Loader
-    REF 26a38983cbe5824fd5be03eab8d037758fc44360
-    SHA512 3029f758ff0c39b57aa10d881af68e73532fd179c54063ed1d4529b7d6e27a5219e3c24b7fb5598d790ebcdc2441e00001a963671dc90fef2fc377c76d724f54
+    REF e6e30ab9c7a61c171cf68d2e7f5c0ce28e2a4eae
+    SHA512 f3563c0a4c094d3795d8386ec0db41189d350ab8136d80ae5de611ee3db87fbb0ab851bad2b33e111eddf135add5dbfef77d96979473ca5a23c036608d443378
     HEAD_REF master
 )
 
+#vcpkg_apply_patches(
```
[the remark this answers is not in the document]
Comment (MathiasMagnus) [reply ports/opencl/portfile.cmake]: OpenCL used to force dynamic linkage. The question is unresolved just yet. There is internal Khronos discussion around supporting/allowing/disallowing it. Current master supports static linkage. (I too believe it is harmless, hence I patched the source scripts but I'll accept consensus and update the package accordingly once the group has reached an agreement.)

File: ports/opencl/portfile.cmake [hunk 3]
```diff
@@ -49,34 +55,66 @@ message(STATUS "Generating OpenCL C++ headers done")
...
+#  set(ICD_SHARED ON)
+#endif()
+
 vcpkg_configure_cmake(
     SOURCE_PATH ${SOURCE_PATH}
     PREFER_NINJA
     OPTIONS
-        -DOPENCL_INCLUDE_DIRS=${CURRENT_PACKAGES_DIR}/include
+        -DOPENCL_ICD_LOADER_HEADERS_DIR=${CURRENT_PACKAGES_DIR}/include
+#        -DBUILD_SHARED_LIBS=${ICD_SHARED}
```
[the remark this answers is not in the document]
Comment (MathiasMagnus) [reply ports/opencl/portfile.cmake]: Tick

File: ports/opencl/portfile.cmake [hunk 4]
```diff
@@ -49,34 +55,66 @@ message(STATUS "Generating OpenCL C++ headers done")
...
-    DESTINATION
-        ${CURRENT_PACKAGES_DIR}/lib
-)
-
-file(INSTALL
-        "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/OpenCL.lib"
-    DESTINATION
-        ${CURRENT_PACKAGES_DIR}/debug/lib
-)
+if(NOT VCPKG_CMAKE_SYSTEM_NAME) # Empty when Windows
```
[the remark this answers is not in the document]
Comment (MathiasMagnus) [reply ports/opencl/portfile.cmake]: Tick

File: ports/opencl/portfile.cmake [hunk 5]
```diff
@@ -49,34 +55,66 @@ message(STATUS "Generating OpenCL C++ headers done")
...
+          ${CURRENT_PACKAGES_DIR}/lib
+  )
+  
+  file(INSTALL
+          "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/OpenCL.lib"
+      DESTINATION
+          ${CURRENT_PACKAGES_DIR}/debug/lib
+  )
+endif()
+if(VCPKG_CMAKE_SYSTEM_NAME STREQUAL "Linux")
```
[the remark this answers is not in the document]
Comment (MathiasMagnus) [reply ports/opencl/portfile.cmake]: Tick

**Comment by MathiasMagnus** [comment]
The WDK requirement (AFAIU) is for older drivers which still load using WDK routines.

>    By default, the OpenCL ICD Loader on Windows requires the Windows Driver Kit (WDK). An OpenCL ICD Loader may be built without the Windows Driver Kit using the CMake variable OPENCL_ICD_LOADER_REQUIRE_WDK, however this option should be used with caution since it may prevent the OpenCL ICD Loader from enumerating some OpenCL implementations. This dependency may be removed in the future.

It may be turned into a feature. I decided not to, because in reality it's an anti-feature that is subject to deprecation. Nontheless I can make it a feature for those running such systems.

**Comment by MathiasMagnus** [comment]
Should fixing missing PDBs with correct usage of `VCPKG_BUILD_TYPE` be made part of this PR? I see my packages also regress on #6368

**Comment by cbezault** [comment]
You can make a separate PR for that.

**Pull request merged**